ctru-rs/src/applets/swkbd.rs

```diff
@@ -99,32 +101,32 @@ impl Swkbd {
     ///
     /// The text received from the keyboard will be truncated if it is greater than 2048 bytes
     /// in length.
-    pub fn get_utf8(&mut self, buf: &mut String) -> Result<Button, Error> {
+    pub fn write_to_string(&mut self) -> Result<(String, Button), Error> {
         // Unfortunately the libctru API doesn't really provide a way to get the exact length
         // of the string that it receieves from the software keyboard. Instead it expects you
         // to pass in a buffer and hope that it's big enough to fit the entire string, so
         // you have to set some upper limit on the potential size of the user's input.
         const MAX_BYTES: usize = 2048;
-        let mut tmp = [0u8; MAX_BYTES];
-        let button = self.get_bytes(&mut tmp)?;
+        let mut tmp = vec![0u8; MAX_BYTES];
+        let button = self.write_bytes(&mut tmp)?;
 
         // libctru does, however, seem to ensure that the buffer will always contain a properly
         // terminated UTF-8 sequence even if the input has to be truncated, so these operations
         // should be safe.
         let len = unsafe { libc::strlen(tmp.as_ptr()) };
-        let utf8 = unsafe { str::from_utf8_unchecked(&tmp[..len]) };
+        tmp.truncate(len);
 
-        // Copy the input into the user's `String`
-        *buf += utf8;
-        Ok(button)
+        let res = unsafe { String::from_utf8_unchecked(tmp) };
+
+        Ok((res, button))
     }
 
     /// Fills the provided buffer with a UTF-8 encoded, NUL-terminated sequence of bytes from
     /// this software keyboard.
     ///
     /// If the buffer is too small to contain the entire sequence received from the keyboard,
-    /// the output will be truncated but should still be well-formed UTF-8
-    pub fn get_bytes(&mut self, buf: &mut [u8]) -> Result<Button, Error> {
+    /// the output will be truncated but should still be well-formed UTF-8.
+    pub fn write_bytes(&mut self, buf: &mut [u8]) -> Result<Button, Error> {
```

Comment:
actually per my other comment maybe [`write_exact`](https://doc.rust-lang.org/std/io/trait.Read.html#method.read_exact) or something like that would a good name.

... now I'm wondering if we should just `impl Read` for this struct, but I guess the `Button` part of the result makes it a little trickier to do that

Reply:
I thought about implementing `Read` already, but the nature of the applet is a bit more complicated than a simple byte read, so I tried to change the naming conventions instead.